compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+                            match_found = True
+                        num_candidates_considered += 1
+                        if num_candidates_considered == self.max_num_matches_considered:
+                            # only consider max_num_matches_considered to limit complexity
+                            break
+                if match_found:
+                    break
+                else:
+                    # if match not found, we index the current substr
+                    self.index_window_upto_pos(match_start_pos + 1)
+
+            if not match_found:
+                # no match found anywhere so put everything else as a literal and break
+                literals += self.window[pos_in_window:]
+                # make sure entire window is indexed
+                self.index_window_upto_pos(len(self.window))
+                break
+            else:
+                # match was found so we appropriately insert into literals and sequences
+                # first put part from pos_in_window to match_start_pos in literals
+                literal_count = match_start_pos - pos_in_window
+                literals += self.window[pos_in_window:match_start_pos]
+                # compute the offset
+                match_offset = match_start_pos - best_match_pos
+                lz77_sequences.append(LZ77Sequence(literal_count, best_match_length, match_offset))
+                match_end_pos = match_start_pos + best_match_length
+                # index the covered portion into the dict
+                self.index_window_upto_pos(match_end_pos)
+                # update position in window
+                pos_in_window = match_end_pos
+
+        return lz77_sequences, literals
+
+    def encode_block(self, data_block: DataBlock):
+        # first do lz77 parsing
+        lz77_sequences, literals = self.lz77_parse_generate_sequences(data_block)
+        # now encode sequences and literals
+        lz77_sequences_encoding = self.encode_lz77_sequences(lz77_sequences)
+        literals_encoding = self.encode_literals(literals)
+        return lz77_sequences_encoding + literals_encoding
+
+    def encode_file(self, input_file_path: str, encoded_file_path: str, block_size: int = 10000):
+        """utility wrapper around the encode function using Uint8FileDataStream
+
+        Args:
+            input_file_path (str): path of the input file
+            encoded_file_path (str): path of the encoded binary file
+            block_size (int): choose the block size to be used to call the encode function
+        """
+        # call the encode function and write to the binary file
+        with Uint8FileDataStream(input_file_path, "rb") as fds:
+            with EncodedBlockWriter(encoded_file_path) as writer:
+                self.encode(fds, block_size=block_size, encode_writer=writer)
+
+
+class LZ77Decoder(DataDecoder):
+    def __init__(self, window_initialization: List = None):
```

Comment:
should this be called `initial_window` instead of `window_initialization`? (minor)

Reply:
done now